### ISSUE TITLE
Extracting APC Back-UPS BX750MI shutdown command from "sniffed" USB traffic

### ISSUE THREAD

**Issue by ErikDB87** [networkupstools/nut#2683]
In light of issue #2666, I "sniffed" USB traffic from APC's `PowerChute Serial Shutdown` software (with the help of this guide (https://www.youtube.com/watch?v=0MC-D_oNzbk)). I hope to be able to find out the shutdown command for my APC "Back-UPS BX750MI FW:295202G -302202G". APC's `PowerChute Serial Shutdown` software is able to shut the UPS down, so there must be a command, is my reasoning.

**Any help on figuring this command out from the pcap files, is welcome!** I haven't got the slightest idea on how to go about it...

I was able to generate two logs. The scenario was both initially the same: I unplugged the UPS from the electricity grid, and the `PowerChute Serial Shutdown` software started closing down my (Windows) laptop. Then I tried two scenarios:

- `2024-11-14-APC-BX750MI-1.pcap`: I let everything run its course, and the UPS shut down, a little over a minute after my laptop had shut down completely (so maybe two minutes after initiating the process, as referenced on that forum I linked somewhere above?). I then plugged the UPS back in, and it restarted (of course).
- `2024-11-14-APC-BX750MI-2.pcap`: As soon as Windows notified me of the shutdown that was going to occur, I plugged the UPS back in. I was pleasantly surprised that the same process as in scenario 1 happened, and the UPS power cycled.

The only caveat is of course that I don't know whether the `PowerChute Serial Shutdown` software shut down `USBPcap` prior to giving its final command - the holy grail - to the UPS... I don't think this is the case: I saw `2024-11-14-APC-BX750MI-2.pcap` "growing" even after I plugged the UPS back in.

(I can't upload the log files, because they're not an allowed file type, so here's a link to a directory on my Google Drive, where they can be found: https://drive.google.com/drive/folders/12xLZH4YWJGXEjNETs2YBJix4boIUVDAE?usp=sharing.)

Thanks in advance to any and all! ;)

**Comment by ErikDB87** [comment]
I made another "sniff" log (`2024-11-19-APC-BX750MI.pcap`, via above mentioned link), this time with the setting that my Windows system would be shut down after 1 second of battery usage, and I unplugged the USB cable as soon as Windows notified me of the shutdown. That way I hoped to keep the log file shorter, and the shutdown command as close to the end of it as possible. The output was still 712 entries long, though...

**Comment by ErikDB87** [comment]
Looks like this is not a very popular approach, and I don't like lingering open issues. :)

**Comment by jimklimov** [comment]
For posterity (if someone with USB programming chops does come around this issue), I'll try to save a copy of your findings as a ZIP file (as fetched from Google Drive today):
[NUT-20250611T082920Z-1-001.zip](https://github.com/user-attachments/files/20687160/NUT-20250611T082920Z-1-001.zip)

**Comment by jimklimov** [comment]
Actually, I think I'll keep this issue open, so it is not completely lost from sight. There are quite a few things to address in NUT backlog, and maybe some future visitor one day will find and fix this.

But thanks for reporting, and for trying to clean up as many other projects like to do :)

**Comment by faluc-dev** [comment]
We decoded the three USBPcap captures from the archived zip (`NUT-20250611T082920Z-1-001.zip`) with a small custom parser (happy to share it). Findings, since nobody seems to have decoded these yet:

**PowerChute Serial Shutdown's entire kill action is a single standard HID write: `SET_REPORT(Feature, ReportID 0x40) = 0x01`** — i.e. `UPS.APCGeneralCollection.APCDelayBeforeReboot` (vendor usage `0xFF86007C`), written exactly once per session, always while the UPS is on battery. A fixed ~2-minute internal grace follows (the value is not a seconds count), then the unit cuts output and re-energizes when AC is present.

Per-capture detail:

- `11-14-1` (full run): `GET_REPORT 0x40` → `40 00`, then `SET_REPORT 0x40 = 40 01` at t=48.3 s on battery; output cut ~2 min later per the original reporter's narration (the capture file itself ends earlier, when PowerChute halts the capturing Windows host).
- `11-14-2`: AC was restored ~5 s after the write — the unit **still completed the power-cycle** ~2 min after the write. The kill is race-immune against wall power returning mid-grace.
- `11-19`: same write at t=12.9 s; capture truncated by the USB yank.

Negative results worth recording: PowerChute never writes `DelayBeforeShutdown` (report 0x15), never touches any report ID ≥ 0x80 (so no hidden vendor byte-pipe/serial tunnel — plain HID feature reports only). The only other write anywhere is Feature `0x11` = current RemainingCapacity at session start and on each capacity change (bookkeeping, not part of the shutdown path). Everything else is a ~1.5 s GET_REPORT polling cycle plus the interrupt stream alternating reports 0x0c/0x16.

**Live confirmation on a different unit** (Back-UPS BX750MI-GR, mfr date 2025-11, usbhid-ups from NUT 2.8.5): `upscmd <ups> shutdown.reboot 1` sent on battery armed `ups.timer.reboot`=1; we re-plugged wall power ~45 s later (mid-grace); output still cut at +2 min 4 s and re-energized 4 s after that — byte-for-byte the captured PowerChute behavior. With the apc-hid **default value ("10") nothing ever fires** — which we believe is precisely why `upsdrvctl shutdown` fails on these units: the built-in shutdown chain reaches `shutdown.reboot`, writes 10, gets USB-level success, and stops. Additionally on this firmware: `load.off.delay` (report 0x15) runs a real countdown but executes only if on battery at expiry (silent no-op on line power), with output auto-returning on AC restore — consistent with the on-battery-only conclusion reached empirically in #2666.

Suggested takeaway for apc-hid: on devices of this family (where `APCDelayBeforeReboot` is the only shutdown-capable usage), the effective `shutdown.reboot` default of "10" is a no-op while "1" is the PowerChute-compatible guaranteed power-cycle. Defaulting this usage's write to 1 (or documenting `sdcommands`/an explicit value for BX-MI units) would make out-of-the-box killpower work on them.

**Comment by jimklimov** [comment]
Cheers, thanks for the analysis - any chance for a pull request about options to change the defaults for BX*MI devices, and conditional command chains to shutdown/restart the UPS (checking if wall power is on/off at the moment, if that matters)?

**Comment by faluc-dev** [comment]
I'd be glad to help in any way, however I need to make something clear first: I am not a coder. I'm running my homelab project in Claude Code, and the analysis above (and the proposal below) was authored by Fable 5. It was monitoring the system while I was pulling and plugging cables. Any and all code in the PR would be written by the AI model. If that's acceptable for a contribution, here's the proposed scope:

The minimal fix looks like one table line in `drivers/apc-hid.c`: `shutdown.return` already maps to `APCDelayBeforeReboot` with value "1" — but only via the `UPS.Output` path, while these BX*MI units expose the usage under `UPS.APCGeneralCollection`, where the only matching entry today is `shutdown.reboot` with the inert "10". Adding

    { "shutdown.return", 0, 0, "UPS.APCGeneralCollection.APCDelayBeforeReboot", NULL, "1", HU_TYPE_CMD, NULL },

means the default shutdown chain (`shutdown.return,shutdown.reboot,load.off.delay,shutdown.stayoff`) succeeds on its very first command with the PowerChute-proven value. I'd verify with a master build and an on-battery drill on our units before opening the PR, and include a NEWS entry.

The conditional-chain question (checking wall power before choosing a command) I'd leave out of this first PR: on this firmware the register itself is the conditional — writes are accepted anytime but executed only while on battery, and the cycle is race-immune if AC returns mid-grace — so an unconditional `shutdown.return` write appears safe in the normal killpower path. The one case we haven't tested is writing the value while on line power throughout: whether it silently drops (like an on-line `DelayBeforeShutdown` expiry does) or arms and fires later. We can bench-test that too and report back, since it decides whether any conditional logic is needed at all.

**Comment by jimklimov** [comment]
Sounds good, quality output from AI is acceptable to move progress forward here, so eagerly awaiting your team's PR :)

**Comment by faluc-dev** [comment]
PR opened: #3566. It also answers the wall-power question from your comment: we tested the same write with AC present throughout — the firmware ACKs it but silently drops it (no execution within 8 minutes, `ups.timer.reboot` reads back disarmed, and a battery transition minutes later produced no delayed fire). So on this family no conditional command chain is needed — the register itself only ever executes on battery, and there it is race-immune. Details and the full test matrix are in the PR.

**Comment by jimklimov** [comment]
Thanks! Just to clarify, is there a way to request the reboot command that works even if AC power *is* present? With NUT, it is common to begin a shutdown because battery is running low, and then the "primary" system ends its uptime by telling the UPS to turn off and come back on when "wall power" returns, or power-cycle and reboot everyone who has shut down if it already returned...

**Comment by faluc-dev** [comment]
As far as we can tell: no — on this firmware there is no way to request a power-cycle while AC is present, and the evidence is fairly complete:

- The device's full HID descriptor exposes exactly two shutdown-capable usages: `UPS.PowerSummary.DelayBeforeShutdown` (0x15) and `UPS.APCGeneralCollection.APCDelayBeforeReboot` (0x40) — the latter is the *only* member of its collection. There is no `DelayBeforeStartup`, no other Output-collection delay control, nothing else writable that touches output state.
- `0x40=1` written with AC present is discarded: no execution within 8 minutes, the register reads back disarmed (`ups.timer.reboot: 0`; armed it holds `1`), and a battery transition several minutes after the write produced no delayed fire.
- `0x15` runs a real countdown regardless of line state, but expiry with AC present is a silent no-op; it only cuts if on battery at expiry.
- Vendor parity: in the captured sessions PowerChute's entire kill is one on-battery `0x40=1` write — nothing else in ~7,900 packets. If APC's own software had an on-line cycle command for these units, it presumably would use it (its use case is the same as upsmon's).

The mitigating detail is that the race window is much narrower than the classic scenario: once the write lands **on battery**, AC returning at *any* later point does not cancel it — the cycle still completes ~2 min after the write and output re-energizes, so every shut-down host gets its AC edge (capture `11-14-2`, plus our live retest with the wall replugged mid-grace). The only uncovered case is AC returning *before* the write lands. And in that case the host still has wall power through the UPS, so the host-side `POWEROFF_WAIT` sleep-then-reboot backstop covers recovery without needing the UPS's cooperation — on this family the answer to the power race seems to be host-side, not UPS-side.

Happy to bench-test any other candidate write you'd like ruled in or out — the unit is on a ready bench setup.

**Comment by faluc-dev** [comment]
We've now bench-tested the second APC in our fleet — a **Back-UPS XS 700U** (fw `924.Z5 .I`, USB fw `Z5`, same `051d:0002` USB id, a ~2022-generation unit) — with the same protocol as the BX750MI work above. Headline: **the value-sensitivity and the fix in PR #3566 apply to this older generation too**, but the surrounding semantics differ enough that they seem worth recording.

Same core situation as the BX:

- The descriptor has no `DelayBeforeStartup` anywhere, and `APCDelayBeforeReboot` (report 0x40, 8-bit) sits under `UPS.APCGeneralCollection` — so the stock chain lands on `shutdown.reboot` = "10", which is ACKed and never executed. **`0x40=1` executes** (bench-proven repeatedly, load observed). The PR's added `shutdown.return` table line would make default `upsdrvctl shutdown` work on this family as well.
- Two extra usages the BX lacks, both unmapped by `apc-hid` and untested by us (no hidraw on the macOS bench): `UPS.APCGeneralCollection.APCDelayBeforeShutdown` (report 0x41, 16-bit) and `UPS.APCGeneralCollection.DelayBeforeShutdown` (report 0x42, 16-bit).

Where the XS diverges from the BX-2025 firmware:

| Behavior | BX750MI (2025 fw) | XS 700U (924.Z5) |
|---|---|---|
| `0x40=1` written on line | silently discarded | **arms** and runs to expiry |
| `0x40=1` grace | ~2 min | **~60 s** |
| 0x40 expiry with AC present | n/a (never armed) | **~0.5 s output cycle** (camera-verified) |
| 0x40 expiry on battery | cut; re-energize on AC return | same (return ~2 s after AC) |
| `0x15` (`load.off.delay`) expiry on line | silent no-op | **cuts** |
| Off-state after a 0x15 cut | auto-return on AC | **sticky** — AC edges do NOT re-energize; only the front-panel button does |

Two practical consequences:

1. **`sdcommands = load.off.delay` advice is firmware-dependent on this family.** On the BX it's the right answer; on the XS it's a stranding hazard — the cut is honored on line *and* on battery, and the resulting off-state survives AC return, so a `load.off.delay` killpower leaves the load dark until someone presses the button. (Battery-*exhaustion* off is different: that state does auto-restart on AC return — observed on this unit in a real outage.) `shutdown.reboot` with value 1 is the only self-recovering kill on the XS, which is another argument for fixing the default chain (the PR) rather than steering users to per-device `sdcommands`.
2. The 0x40 race behavior differs in mechanism but not in outcome: on the XS the armed timer is not cancelled by AC return — expiry with AC present is a ~0.5 s output cycle, on battery it's a cut-until-AC. Either way every shut-down host eventually gets an AC edge, same net effect as the BX's race-immune cycle.

Bench-method notes for anyone testing these units:

- `ups.status` is unreliable around timer operations (`OL DISCHRG` on line during countdowns, `OB`/`OL` blips both ways, `OL` reported while on battery in the off-state), and `input.voltage` holds a stale reading after AC loss (decays via ~140 V). Our ground truth was a filament desk lamp filmed with a phone camera: the eye sees millisecond transfer-relay dips the 30/60 fps video integrates over, so "dark frames on video" cleanly separates a real output cycle (the 0.5 s event above) from relay transfer clicks.
- The XS's USB iSerial carries two trailing spaces (`"3B2022X70450  "`) — a `serial` matcher in `ups.conf` needs to account for them.

Full HID walk and per-second poller logs from the bench are archived; happy to run any other write you'd like ruled in or out on either unit.
